n3fit/src/n3fit/model_gen.py

```diff
@@ -323,84 +323,202 @@ def observable_generator(
     return layer_info
 
 
+@dataclass
+class ReplicaSettings:
+    """Dataclass which hold all necessary replica-dependent information of a PDF.
+
+    Parameters
+    ----------
+        seed: int
+            seed for the initialization of the neural network
+        nodes: list[int]
+            nodes of each of the layers, starting at the first hidden layer
+        activations: list[str]
+            list of activation functions, should be of equal length as nodes
+        architecture: str
+            select the architecture of the neural network used for the replica,
+            e.g. ``dense`` or ``dense_per_flavour``
+        initializer: str
+            initializer to be used for this replica
+        dropout: float
+            rate of dropout for each layer
+        regularizer: str
+            name of the regularizer to use for this replica (if any)
+        regularizer_args: dict
+            options to pass down to the regularizer (if any)
+    """
+
+    seed: int
+    nodes: list[int]
+    activations: list[str]
+    architecture: str = "dense"
+    initializer: str = "glorot_normal"
+    dropout_rate: float = 0.0
+    regularizer: str = None
+    regularizer_args: dict = field(default_factory=dict)
+
+    def __post_init__(self):
+        """Apply checks to the input, and expand hyperopt callables"""
+        # Expansions
+        if callable(self.activations):
+            # Hyperopt might pass down a function to generate the list of activations
+            # depending on the number of layers
+            self.activations = self.activations(len(self.nodes))
+
+        if self.regularizer_args is None:
+            self.regularizer_args = dict()
+
+        # Checks
+        if len(self.nodes) != len(self.activations):
+            raise ValueError(
+                f"nodes and activations do not match ({self.nodes} vs {self.activations}"
+            )
+        if self.regularizer_args and self.regularizer is None:
+            raise ValueError(
+                "Regularizer arguments have been provided but not regularizer is selected"
+            )
+
+
+def sample_pdf_models():
+    """
+    Given a set of parameters, sample a set of PDF models
+    """
+
+
 def generate_pdf_model(
-    nodes: list[int] = None,
-    activations: list[str] = None,
-    initializer_name: str = "glorot_normal",
-    layer_type: str = "dense",
+    replicas_settings: list[ReplicaSettings],
     flav_info: dict = None,
     fitbasis: str = "NN31IC",
     out: int = 14,
-    seed: int = None,
-    dropout: float = 0.0,
-    regularizer: str = None,
-    regularizer_args: dict = None,
     impose_sumrule: str = None,
     scaler: Callable = None,
-    num_replicas: int = 1,
     photons: Photon = None,
```

Comment:
Just to say that in the (near) future, the arguments of this function will contain `nuclear_info` which will contain the information about the atomic mass number $A$ of the hadron. Then the input of the NN will become `(x, log(x)/scaler(x), A)`.

Reply:
I think that can be at the same level as photon. I was thinking (but let's leave it to another PR) to add a `nuclear` and `fragmentation` namespaces which will arrive here together with photon and include their own replica settings (which I don't think we want to sample as the precision is not there yet)

But I still don't know whether it is the best option